xfer/pipes.go

```diff
@@ -0,0 +1,143 @@
+package xfer
+
+import (
+	"io"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+// Pipe is a bi-directional channel from someone thing in the probe
+// to the UI.
+type Pipe interface {
+	ID() string
+	Port() io.ReadWriter
+	Starboard() io.ReadWriter
+	CopyToWebsocket(io.ReadWriter, *websocket.Conn) error
+
+	Close(time.Time)
+	Closed() bool
+	ClosedAt() time.Time
+	OnClose(func())
+}
+
+type pipe struct {
+	mtx             sync.Mutex
+	id              string
+	port, starboard io.ReadWriter
+	tombstone       time.Time
+	quit            chan struct{}
+	onClose         func()
+}
+
+// NewPipe makes a new... pipe.
+func NewPipe(id string) Pipe {
+	r1, w1 := io.Pipe()
+	r2, w2 := io.Pipe()
+	return &pipe{
+		id: id,
+		port: struct {
+			io.Reader
+			io.Writer
+		}{
+			r1, w2,
+		},
+		starboard: struct {
+			io.Reader
+			io.Writer
+		}{
+			r2, w1,
+		},
+		quit: make(chan struct{}),
+	}
+}
+
+func (p *pipe) ID() string {
+	return p.id
+}
+
+func (p *pipe) Port() io.ReadWriter {
+	return p.port
+}
+
+func (p *pipe) Starboard() io.ReadWriter {
+	return p.starboard
+}
+
+func (p *pipe) Close(t time.Time) {
```

Comment:
I was expecting that calling close on either end would close the entire thing.

I was using `tombstoneTime` to refer to the duration that the tombstone would exist for, but looking now, that's never referenced by the pipe, so wouldn't need to be passed in.

Again, not sure it's a good idea, but my hunch is that it seems like it *should* simplify usage.

Total strawman:
```Go
func duplex(a, b io.ReadWriteCloser, cancel <-chan struct{}) error {
  // This implementation is totally made up
  errs := make(chan error)

  // close everything when one exists
  defer a.Close()
  defer b.Close()

  // copy everything
  go func() { errs <- io.Copy(a, b) }()
  go func() { errs <- io.Copy(b, a) }()

  // wait for one to exit
  select {
  case err := <-errs:
    return err
  case <-cancel:
    return nil
  }
}

// in the probe, maybe:
cancel := make(chan struct{})
p := NewPipe()
local, remote := p.Ends()
go duplex(local, docker, cancel)
go duplex(remote, websocket, cancel)
p.Wait() // Wait for pipe to be closed. Pipe is closed when either end is closed.
```

Maybe that won't work, though...